python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,471 @@
+        return True
+
+    def satisfies_order_of(self, other: Transform) -> bool:
+        return other.preserves_order()
+
+
+class Truncate(Transform):
+    """A transform for truncating a value to a specified width.
+
+    Args:
+      source_type (Type): An Iceberg Type of IntegerType, LongType, StringType, or BinaryType
+      width (int): The truncate width
+
+    Raises:
+      ValueError: If a type is provided that is incompatible with a Truncate transform
+    """
+
+    _VALID_TYPES = {IntegerType, LongType, StringType, BinaryType}
+    _TO_HUMAN_STR = {BinaryType: transform_util.base64encode}
+
+    def __init__(self, source_type: Type, width: int):
+        if source_type not in Truncate._VALID_TYPES and not isinstance(
+            source_type, DecimalType
+        ):
+            raise ValueError(f"Cannot truncate type: {source_type}")
+
+        super().__init__(
+            f"truncate[{width}]",
+            f"transforms.truncate(source_type={repr(source_type)}, width={width})",
+            Truncate._TO_HUMAN_STR.get(source_type, str),
+        )
+        self._type = source_type
+        self._width = width
+
+        if self._type == IntegerType or self._type == LongType:
+            self._apply = lambda v, w: v - v % w
+        elif self._type == StringType or self._type == BinaryType:
+            self._apply = lambda v, w: v[0 : min(w, len(v))]
+        else:  # decimal case
+            self._apply = transform_util.truncate_decimal
+
+    @property
+    def width(self):
+        return self._width
+
+    def apply(self, value):
+        if value is None:
+            return None
+        return self._apply(value, self._width)
+
+    def can_transform(self, target: Type) -> bool:
+        return self._type == target
+
+    def preserves_order(self) -> bool:
+        return True
+
+    def satisfies_order_of(self, other: Transform) -> bool:
+        if self == other:
+            return True
+        elif (
+            StringType == self._type
+            and isinstance(other, Truncate)
+            and StringType == other._type
+        ):
+            return self._width >= other._width
+
+        return False
+
+
+class UnknownTransform(Transform):
+    """A transform that represents when an unknown transform is provided
+
+    Args:
+      source_type (Type): An Iceberg `Type`
+      transform (str): A string name of a transform
+
+    Raises:
+      AttributeError: If the apply method is called.
+    """
+
+    def __init__(self, source_type: Type, transform: str):
+        super().__init__(
+            transform,
+            f"UnknownTransform(source_type={repr(source_type)}, transform={repr(transform)})",
+        )
+        self._type = source_type
+        self._transform = transform
+
+    def apply(self, value):
+        raise AttributeError(f"Cannot apply unsupported transform: {self.__str__()}")
+
+    def can_transform(self, target: Type) -> bool:
+        return repr(self._type) == repr(target)
```

Comment:
I'm really skeptical that comparing a `repr` is the right thing to do.

Reply:
Yeah, did some investigation and looks like Type classes do not implement `__eq__`. As this PR is quite large, I will send another PR to add `__eq__` to types. I will update here to be `self._type == target`